merkledag.go

```diff
@@ -0,0 +1,88 @@
+package format
+
+import (
+	"context"
+	"fmt"
+
+	cid "github.com/ipfs/go-cid"
+)
+
+var ErrNotFound = fmt.Errorf("merkledag: not found")
+
+// Either a node or an error.
+type NodeOption struct {
+	Node Node
+	Err  error
+}
+
+// TODO: This name kind of sucks.
+// NodeResolver?
+// NodeService?
+// Just Resolver?
+type NodeGetter interface {
+	Get(context.Context, *cid.Cid) (Node, error)
+}
+
+// DAGService is an IPFS Merkle DAG service.
+type DAGService interface {
+	NodeGetter
+
+	Add(Node) (*cid.Cid, error)
+	Remove(Node) error
+
+	// TODO: This is returning them in-order?? Why not just use []NodePromise?
+	// Maybe add a couple of helpers for getting them in-order and as-available?
+	// GetDAG returns, in order, all the single leve child
+	// nodes of the passed in node.
+	GetMany(context.Context, []*cid.Cid) <-chan *NodeOption
+
+	Batch() Batch
+
+	LinkService
+}
+
+// An interface for batch-adding nodes to a DAG.
+
+// TODO: Is this really the *right* level to do this at?
```

Comment:
The reason for this abstraction is to avoid forcing callers to think about batching. This batch can determine the right time to flush on its own for optimal memory vs batch size tradeoffs, if done differently, each caller would have to reimplement similar logic, and that increases the chance of them doing something incorrectly